ouroboros-consensus-cardano-tools/src/Cardano/Tools/ImmDBServer/Diffusion.hs

```diff
@@ -0,0 +1,115 @@
+import           Network.Socket (SockAddr (..))
+
+import           Ouroboros.Network.ErrorPolicy (nullErrorPolicies)
+import           Ouroboros.Network.IOManager (withIOManager)
+import           Ouroboros.Network.Mux
+import qualified Ouroboros.Network.NodeToNode as N2N
+import qualified Ouroboros.Network.Snocket as Snocket
+import           Ouroboros.Network.Socket (configureSocket)
+
+import           Cardano.Tools.ImmDBServer.MiniProtocols (immDBServer)
+import           Ouroboros.Consensus.Block
+import           Ouroboros.Consensus.Config
+import           Ouroboros.Consensus.Config.SupportsNode
+import           Ouroboros.Consensus.Node.InitStorage
+                     (NodeInitStorage (nodeCheckIntegrity, nodeImmutableDbChunkInfo))
+import           Ouroboros.Consensus.Node.NetworkProtocolVersion
+import           Ouroboros.Consensus.Node.Run (SerialiseNodeToNodeConstraints)
+import           Ouroboros.Consensus.Storage.FS.API (SomeHasFS (..))
+import           Ouroboros.Consensus.Storage.FS.API.Types
+                     (MountPoint (MountPoint))
+import           Ouroboros.Consensus.Storage.FS.IO (ioHasFS)
+import qualified Ouroboros.Consensus.Storage.ImmutableDB as ImmutableDB
+import           Ouroboros.Consensus.Util
+import           Ouroboros.Consensus.Util.IOLike
+import           Ouroboros.Consensus.Util.ResourceRegistry
+
+serve ::
+     SockAddr
+  -> N2N.Versions N2N.NodeToNodeVersion N2N.NodeToNodeVersionData
+       (OuroborosApplication 'ResponderMode SockAddr BL.ByteString IO Void ())
+  -> IO Void
+serve sockAddr application = evalContT do
+    iocp <- ContT withIOManager
+    let sn     = Snocket.socketSnocket iocp
+        family = Snocket.addrFamily sn sockAddr
+    socket <- ContT $ bracket (Snocket.open sn family) (Snocket.close sn)
+    lift do
+      networkMutableState <- N2N.newNetworkMutableState
+      configureSocket socket (Just sockAddr)
+      Snocket.bind sn socket sockAddr
+      Snocket.listen sn socket
+      N2N.withServer
+        sn
+        N2N.nullNetworkServerTracers {
+          N2N.nstHandshakeTracer   = show >$< stdoutTracer
+        , N2N.nstErrorPolicyTracer = show >$< stdoutTracer
+        }
+        networkMutableState
+        acceptedConnectionsLimit
+        socket
+        application
+        nullErrorPolicies
+  where
+    acceptedConnectionsLimit = N2N.AcceptedConnectionsLimit {..}
+      where
+        acceptedConnectionsHardLimit = maxBound
+        acceptedConnectionsSoftLimit = maxBound
+        acceptedConnectionsDelay     = 0
+
+run ::
+     forall blk.
+     ( GetPrevHash blk
+     , ShowProxy blk
+     , SupportedNetworkProtocolVersion blk
+     , SerialiseNodeToNodeConstraints blk
+     , ImmutableDB.ImmutableDbSerialiseConstraints blk
+     , NodeInitStorage blk
+     , ConfigSupportsNode blk
+     )
+  => FilePath
+  -> SockAddr
+  -> TopLevelConfig blk
+  -> IO Void
+run immDBDir sockAddr cfg = evalContT do
+    registry <- ContT withRegistry
```

Comment:
Suggestion: Same as above.

Reply:
Also, something to agree on in coding conventions?

Reply:
See https://github.com/input-output-hk/ouroboros-network/pull/4339#discussion_r1098788510, I think there is a place both for "direct" style as well as `ContT`-style depending on context. :+1:

Reply:
I agree, it's just that I haven't seen any use of `ContT` so far in the whole ouroboros codebase. I have not seen everything for sure :)